File: pldi21_benchmarks/callback_.c
#include <caml/mlvalues.h>
#include <caml/callback.h>

value ocaml_to_c (value v) {
  int n = Int_val(v);
  value c_to_caml_closure = *caml_named_value("c_to_ocaml");
  for (int i = 0; i < n; i++) {
    caml_callback(c_to_caml_closure, Val_unit);
  }
  return Val_unit;
}

File: pldi21_benchmarks/callback.ml
let n = try int_of_string (Sys.argv.(1)) with _ -> 1_000_000
external ocaml_to_c : int -> unit = "ocaml_to_c"

let c_to_ocaml () = ()
let _ = Callback.register "c_to_ocaml" c_to_ocaml

let _ = ocaml_to_c n
